### DISCUSSION TITLE
Pull Analytics: PR turnaround time, lead time for code review and changes

### DISCUSSION THREAD

**Discussion by thibaudcolas** [community/community#13037] [Pull Requests]
I’m surprised there is no discussion thread for this already! I never got to try [Pull Panda’s Pull Analytics](https://web.archive.org/web/20200513115716/https://pullpanda.com/analytics), but from the marketing page this seems like invaluable functionality to keep the pull request review funnel in check.

In the case of [Wagtail](https://github.com/wagtail/wagtail) in particular, we pay close attention to how fast we are at acknowledging the work of new contributors submitting PRs, providing code reviews, and (hopefully) eventually getting their code merged. We see this review turnaround time as one of the most important aspects of the project’s health. 

We [keep track](https://github.com/wagtail/wagtail/wiki/PR-triage) of the various bottlenecks things currently get stuck at with labels and filters, but this is very ad-hoc. It’d be lovely to have actual charts, either auto-generated like the current repository-level Insights, or with manual creation like the [Project Insights](https://github.blog/changelog/2021-10-27-the-new-github-issues-public-beta/) charts. We also keep track of different related metrics ("number of new contributors in last week’s PRs, proportion of last week’s PRs we’ve replied to") in a spreadsheet we manually update on a weekly basis.

## In the meantime, with Project Insights

In the meantime, I think we might be able to get this already if we created a private organization project, and added all pull requests to this project with a different status per step of the funnel we want to track. We could then use the Project Insights burn-down chart to see how many PRs are at different stages (but not how long a given PR has spent at each stage – [mentioned here too](https://github.com/github/feedback/discussions/10915#discussioncomment-2373669)).

This would look something like this:

![project-insights-pr-review-funnel](https://user-images.githubusercontent.com/877585/159090796-bfdb0411-6934-4520-a8fa-a6f3a0e48503.png)

The chart is useful, and being able to see all of the PRs across the different stages all on the same table view is also invaluable. However this requires a fair bit of work to move the PRs between steps to match their label.

## References

Aside from Project Insights, I can see two related items in GitHub’s public roadmap

- On [Insights: Engineering Leader and Organizational DORA Reports #127](https://github.com/github/roadmap/issues/127), the "Lead time for changes" chart. In our case, this would be more about the PR review turnaround time – but close enough.
- On [Insights: Fully integrated reports & dashboards for GHAE #171](https://github.com/github/roadmap/issues/171) – general pull request health metrics.

**Comment by thibaudcolas** [comment]
For anyone interested in this, two relevant things I stumbled upon:

- Another Discussions thread asking for feedback on pull request analytics: #14270
- A third-party project implementing the "Average time to close pull requests": https://www.repotrends.com/

**Reply by hbd** [reply]
+1 this would be very helpful for measuring improvements in review process within a project. 

In addition, would be great to get an average time taken to respond to a review request on an individual level. This is similar to a "usually response in X minutes" in an Intercom chat. For example, "hbd's average response time for a PR request is X minutes". Exactly what a response is (first response after initial review request?), is up for debate.

**Comment by abhishekbedi1432** [comment]
Thanks everyone for inputs. 
I was wondering if there is any Github Action or Tool where we can see the Lead Time ( PR Review time / pipeline action time / etc) on the PR itself to see how fast our CI pipeline is performing.

**Reply by c12k** [reply]
I experimented with looking through commit history, but with pr squash and merge and branch deletion you lose the history time when a brach is created from main ( the key date for tracking time branch started to compare against merge time).

**Comment by sergiinvestor** [comment]
Understanding PR turnaround time and lead time for code reviews is crucial for maintaining project health. In my experience, tools like [Gitlights](https://github.com/marketplace/gitlights-analytics) offer detailed analytics on these metrics, providing insights into review times and identifying potential bottlenecks.
Integrating such tools can enhance your ability to monitor and improve code review processes.

**Comment by jmgarnier** [comment]
Datadog provides this now https://docs.datadoghq.com/dora_metrics/calculation/#change-lead-time-stages
